Design for exponential backoff

irbekrm opened the pull request:
Design for expontial backoff for retrying failed certificate issuances.

Note: this design currently assumes binary exponential backoff with delay periods 1h, 2h, 4h etc up to 32h at which point it would remain at 32h till reset (resetting can only happen via a successful issuance). The delay periods could be changed and that would not affect this design.



Related issues:
- #4727 

Signed-off-by: irbekrm <irbekrm@gmail.com>

irbekrm commented:
/release-note-none

jahrlin commented on design/2022.exponential-backoff.md:
```diff
@@ -0,0 +1,126 @@
+
+### Non-goals
+
+- Make the backoff period configurable (extra complexity). For context Kubernetes pod backoff crashloopbackoff period is _not_ configurable (although it is a very demanded feature [k/k#57291](https://github.com/kubernetes/kubernetes/issues/57291))
```
Very strongly agree!

jahrlin commented on design/2022.exponential-backoff.md:
```diff
@@ -0,0 +1,126 @@
+- Make the backoff period configurable (extra complexity). For context Kubernetes pod backoff crashloopbackoff period is _not_ configurable (although it is a very demanded feature [k/k#57291](https://github.com/kubernetes/kubernetes/issues/57291))
+
+### Must not
+- Break the current abstractions of Certificate, CertificateRequest, Order etc. For example,`CertificateRequest` represents a single attempt to issue a certificate that either succeeded or not if it's in terminal state
```
does this mean "Break compatability", i.e. it must not break compatability, up- and downgrading can be done without extra steps.

irbekrm replied:
No, this is more abstract, an example breaking of abstraction would be, I think, making `CertificateRequest` somehow 'aware' of the next or previous issuances, for example a field that would somehow encode the delay for this particular issuance. It's related to what could be alternative ways to do this, but maybe actually doesn't make that much sense in the context of this doc

jahrlin replied:
Thanks for clarifying. 

**Should** we have a "must not" item specifically for compatibility, or is that implied?

irbekrm replied:
Thanks, it's a very good point 👍🏼 
I've added an 'Upgrading' section that should cover this

jahrlin reviewed (commented):
This is great! Thanks @irbekrm 

I immediately confused this with introducing retries and exponential backoffs for HTTP and gRPC requests made in the cert-manager code.

Can we add a note/distinction in the text and maybe rename the file to reflect that this is specifically for Certificate Issuance?

irbekrm commented on design/2022.exponential-backoff.md:
```diff
@@ -0,0 +1,126 @@
+- Ensure that failed certificate issuances does not overwhelm external services (to prevent putting extra strain on the external services as well as to limit hitting rate limits for cert issuance) by backing off with an exponentially increasing time period. 
+- Ensure that users can see somewhere when the issuance will be retried and that the backoff mechanism does not introduce extra complexity for debugging and fits in with the already existing 1 hour backoffs
+- Ensure that the backoff mechanism is similar to how it is implemented elsewhere on Kubernetes (i.e pod crashloopbackoff)
+- Ensure that cmctl renew can still be used to manually force renewal _now_
```
> The goal here would be to provide an escape hatch for users, and renew would be how to reach that - is that fair to say?

Yes and also this is how `cmctl renew` works now- you can use it to retry renewal immediately without waiting for 1 hour and I have seen that it is being used in cases where the issuance failed then some external fix was made (i.e DNS setup fixed) and the user would like to see if the fix worked without waiting. In case of the delay period being even longer than an hour, it would be even more important to have this escape hatch.
I'll add a note about that 👍🏼

irbekrm commented on design/2022.exponential-backoff.md:
```diff
@@ -0,0 +1,126 @@
+### Implementation
+
+The creation of `CertificateRequest`s for a particular revision of a `Certificate` will happen with a backoff.
+A new `IssuanceAttemptsForRevision` status fields will be added to `Certificate` that will be used to record the number of failed issuances for this revision.
```
Agree that this name is long and ugly.
I did feel like there would be a value in the name somehow suggesting that these are retries _for this revision_ . However, perhaps that information could be better put into a field description.
I will change it to `Retries` for now 👍🏼

jahrlin commented:
/approve
/lgtm
/hold

munnerz commented on design/20220118.certificate-issuance-exponential-backoff.md:
```diff
@@ -0,0 +1,144 @@
+A new `IssuanceAttempts` status field will be added to `Certificate` that will be used to record the number of failed issuances for this revision.
+Similarly to `status.LastFailureTime`, `status.IssuanceAttempts` field will only be set for a currently failing `Certificate` and will be removed after a successful issuance.
+
+`IssuanceAttempts` will be set by `certificates-issuing` controller after a failed issuance by either bumping the already existing value by 1 or setting it to 1 (first failure). In case of a succeeded issuance, `certificates-issuing` controller will ensure that `status.IssuanceAttempts` is not set.
```
Would `failedIssuanceAttempts` be clearer? Personally I always get confused with `issuanceAttempts` sort of wording as it doesn't make it clear whether 1 is the 'normal' value (i.e. one attempt is in progress) or if means failure. `failedIssuanceAttempts` seems more precise and I don't _think_ introduces undue complexity?

irbekrm replied:
I guess the _precise_ wording would be `failedIssuanceAttemptsForThisRevision` :) There was another discussion about the naming of this field already https://github.com/jetstack/cert-manager/pull/4757#discussion_r786635212
I can see how `failedIssuanceAttempts` makes more sense than `issuanceAttempts` logically, but it is longer and still isn't _exactly_ precise :)
What about `issuancesAttempted`?

munnerz commented on design/20220118.certificate-issuance-exponential-backoff.md:
```diff
@@ -0,0 +1,144 @@
+
+## Goals
+
+- Ensure that retrying failed issuances does not overwhelm external services and is less likely to cause issuance being rate-limited by backing off with an exponentially increasing time period. 
```
a bit of a nit: I think it may be better to change "overwhelm external services" to more like "not use up all of a users quota" or something similar - I imagine this *isn't* the feature that prevents a DoS (i.e. huge numbers of HTTP requests) - rather it slows down the rate at which we start a whole new issuance flow.

I only call it out because there's other work going on that *does* aim to prevent overwhelming external services too - and this serves a subtly different (but equally important) area

irbekrm replied:
> a bit of a nit: I think it may be better to change "overwhelm external services" to more like "not use up all of a users quota" or something similar - I imagine this isn't the feature that prevents a DoS (i.e. huge numbers of HTTP requests) - rather it slows down the rate at which we start a whole new issuance flow.

This is a point where I sometimes struggle with imagining how these things are likely to interact in large scale deployments. For a small number of certs from user's perspective retrying every hour should be ok and the problem would likely be only exceeding a quota as you say.
But for a very large installation where something happens such that all cert issuances are suddenly failing and are now all being retried every hour, would the number of HTTP requests also not become an issue?
Similarly, assuming point of view of an external service (i.e Let'sEncrypt) that has a large number of cert-manager using clients- this should reduce the number of calls they receive _overall_ that are less likely to succeed

munnerz commented on design/20220118.certificate-issuance-exponential-backoff.md:
```diff
@@ -0,0 +1,144 @@
+
+- Ensure that retrying failed issuances does not overwhelm external services and is less likely to cause issuance being rate-limited by backing off with an exponentially increasing time period. 
+
+- Ensure that when the backoff is being applied, users have a way to find out when the issuance will be next retried and that the backoff mechanism does not introduce extra complexity for debugging and fits in with the already existing 1 hour backoffs
```
Hm - I recall you talked about setting `nextIssuanceTime` to the next time the renewal attempt will be made on Slack and we eventually thought it'd be best to not do that.

I *don't* see that discussed in the proposal, but this is still a goal (and I think an un-met one with this proposal as-is)? Unless I missed something 😄

irbekrm replied:
I still wanted to give users some way how to tell when the next issuance will be attempted (this type of question seems to appear a lot, so it seems like something that would be useful for their debugging) beyond just telling them to calculate it.

For this design, I was aiming to fulfill that goal with just an event on the `Certificate`, I will add that to the spec 👍🏼 (I am still not entirely sure what would be the best way to expose this to users, but I guess a status field could always be add later if there was a  need for it)

irbekrm replied:
As I was working on the implementation, the event didn't seem all that useful- it is easy to have multiple of these events with different timestamps at the same time time on one failing `Certificate` when manually retrying issuances which is confusing, also the fact that the events are short lived may not help much with debugging.

munnerz commented on design/20220118.certificate-issuance-exponential-backoff.md:
```diff
@@ -0,0 +1,144 @@
+
+## Non-goals
+
+- Make the backoff period configurable (extra complexity). For context Kubernetes pod backoff crashloopbackoff period is _not_ configurable (although it is a very demanded feature [k/k#57291](https://github.com/kubernetes/kubernetes/issues/57291))
```
👍 though I do think we should expose configuration options for this in future personally

irbekrm replied:
It'll probably become more obvious what is needed after we see how this functionality actually works for users

munnerz commented on design/20220118.certificate-issuance-exponential-backoff.md:
```diff
@@ -0,0 +1,144 @@
+
+- Make the backoff period configurable (extra complexity). For context Kubernetes pod backoff crashloopbackoff period is _not_ configurable (although it is a very demanded feature [k/k#57291](https://github.com/kubernetes/kubernetes/issues/57291))
+
+- Make it possible to reset the backoff period _without_ having a successful issuance first (i.e by manually running `cmctl renew`)
```
To be clear, it *is* possible to trigger an immediate *attempt*, but at the end of that attempt the counter *won't* be reset to zero (unless it succeeded), therefore the next issuance after that will have a greater back-off applied, right?

munnerz reviewed (commented):
Thanks for writing this up - this is really detailed and clear as to what you're trying to achieve!

I've added a few clarifying comments but largely LGTM.

My one question is are we comfortable to put this feature straight to 'GA'? I personally think it's okay to do as it's a fairly small extension in controller code vs what we do today, and I don't *think* we'll want to change the API surface we're adding here either (it's pretty simple all round). What do you think? :)

irbekrm commented:
> I've added a few clarifying comments but largely LGTM.

Thanks James, I've made some changes in response to the comments and restructured a doc a bit

> My one question is are we comfortable to put this feature straight to 'GA'? I personally think it's okay to do as it's a fairly small extension in controller code vs what we do today, and I don't think we'll want to change the API surface we're adding here either (it's pretty simple all round)

I also think it could go straight to GA by looking at the small amount of change this actually implies.

The bit that I imagine _might_ change would be the delay periods and how they are calculated. I don't think that it is worth putting this behind a feature gate just to stabilize the delay periods, but perhaps there is a reason to do that.

irbekrm commented:
After having worked on the implementation I've added a couple notes to the design:
- removed references to revision number as the backoff was implemented for all consecutive failures
- added a note that denied `CertificateRequest`s will result in backoff
- removed reference to the event with the next retry timestamp as that didn't seem all that useful

irbekrm commented:
/test pull-cert-manager-e2e-v1-23

irbekrm commented:
Just looking to trigger some tests for unrelated purpose

/test pull-cert-manager-e2e-v1-23

irbekrm commented:
/test pull-cert-manager-e2e-v1-23

irbekrm commented:
/test pull-cert-manager-e2e-v1-23

irbekrm commented:
/test

irbekrm commented:
/retest

irbekrm commented:
/test

irbekrm commented:
/test

maelvls commented:
Hi! As part of the effort of migrating from Bazel to Make, we [have added](https://github.com/jetstack/testing/pull/648) two new Prow jobs that require this PR to be rebased on top of master in order to pass. The two Prow jobs that were added are `pull-cert-manager-make-test` and `pull-cert-manager-make-e2e-v1-23`. Thanks! 🙏

irbekrm commented:
Flake

/test pull-cert-manager-make-test

irbekrm commented:
/milestone v1.8

SgtCoDFish commented:
/lgtm
/unhold

irbekrm commented:
e2e tests failing for a design doc makes perfect sense

irbekrm commented:
/test pull-cert-manager-make-e2e-v1-23

SgtCoDFish commented:
> e2e tests failing for a design doc makes perfect sense

I think it a dependency for the test failed to download :thinking:
